benchmark_test/benchmark_test.go

```diff
@@ -0,0 +1,160 @@
+package benchmark_test
```

Comment:
first: all login in `main` is reused here (just moved out of `main` which is a good idea any case)

second: jenkins will run this as `ENV_VAR_X=X... go test -v ....`, which means that to add a workload we just have to add a unit test; and you can run it locally in the same way
the other way would mean that jenkins would run tests like:
```
./hey-apm benchmark -run 3m -tf 100ms -e 100000 -mx 10 -ms 10 -ef 20 (...many more args)
./hey-apm benchmark -run 3m -tf 100ms -e 0 -mx 30 -ms 30 (...many more args)
...many more lines
```
adding more workloads means adding more lines to some script, which I don't know if I can run locally (jenkins files?), etc.; and it is much more unreadable (like 10 lines with 15 args each), easier to slip typos, etc.

tl;dr golang is more expressive than a make script or similar. i can change it if this has any drawbacks i didn't think about.

Reply:
I was referring to the already existing logic of [parsing user input](https://github.com/elastic/hey-apm/pull/116/files#diff-7ddfb3e035b42cd70649cc33393fe32cR16). Why introducing a second way to parse ENV variables instead of unifying and reusing?

I don't think the comparison of how you would start the tests and how you'd start a benchmark test is fair. You could still trigger _one_ `hey-apm benchmark` with the same options that are provided for starting the testsuite and then this `benchmark` task takes care of starting all the defined use cases. 

Not sure why you are referring to _make_ scripts, I didn't mean or imply to use any _make_ scripts. What I had in mind is something like:

```
struct Benchmark{
  args models.Input
  cases []models.Input
}

func (b *Benchmark) NewBenchmark(i models.Input){ 
  b.args = i
  //define and create cases based on i
}

func (b *Benchmark) Run(){
  for c := range b.cases{
    setup //warmup, cooldown etc
    report := worker.Run(c)
    if err := verify(report){ raise... }
  }
}
```

As I tried to explain in my first comment, I cannot follow your reasons for putting this logic into tests. This is a dedicated repository for running a high load against an external resource and gathering feedback about it. The benchmarking package adds logic to compare those results with past results. 
If you change it or not is up to you, but I'd appreciate if you could explain why you'd rather put this into tests than into regular go logic.

Reply:
Ok, so no make scripts, I misunderstood that, then.

So what you are arguing is that instead of `SERVER_URL=x... go test -v`, this should be executed like `go benchmark -server_url=x...`?

All the extra logic I am introducing is a `getEnv` function (not crazy complicated), and I gain access to all the test assertion functions, which are convenient. 
(and env vars might still be better either case for secrets, security wise)

Since this _tests_ the apm-server for regressions and will run on Jenkins, I also feel is quite natural to define it as tests.
So I still don't see clear downsides of this and advantages of the other, but is not big deal to change it... There are many ways to skin a cat, I guess.

Reply:
My concerns are _conceptually_, not about how the functionality is called.

As you said, many different ways of solving something. Fine with me to leave it as is.